packages/logging_cloudwatch/amplify_logging_cloudwatch/lib/src/queued_item_store/index_db/indexed_db_adapter.dart

```diff
@@ -0,0 +1,177 @@
+/// {@endtemplate}
+class IndexedDbAdapter implements QueuedItemStore {
+  /// {@macro amplify_analytics_pinpoint_dart.indexed_db_adapter}
+  IndexedDbAdapter();
+
+  /// The name of the database.
+  ///
+  /// Reference: https://www.w3.org/TR/IndexedDB/#name
+  String get databaseName => 'amplify_logging_cloudwatch';
+
+  /// The name of the object store.
+  ///
+  /// Reference: https://www.w3.org/TR/IndexedDB/#object-store-name
+  final storeName = 'logging_cached_logs';
+
+  late final Future<void> _databaseOpenEvent = _openDatabase();
+
+  late final IDBDatabase _database;
+
+  int _currentTotalByteSize = 0;
+
+  /// Checks for IDB availability and attempts to open the database.
+  Future<void> _openDatabase() async {
+    final db = indexedDB;
+    if (db == null) {
+      throw const InvalidStateException('IndexedDB is not available');
+    }
+    final openRequest = db.open(databaseName, 1)
+      ..onupgradeneeded = (event) {
+        final database = event.target.result;
+        final objectStoreNames = database.objectStoreNames;
+        if (!objectStoreNames.contains(storeName)) {
+          database.createObjectStore(
+            storeName,
+            keyPath: 'id',
+            autoIncrement: true,
+          );
+        }
+      };
+    _database = await openRequest.future;
+  }
+
+  /// Returns a new [IDBObjectStore] instance after waiting for initialization
+  /// to complete.
+  IDBObjectStore _getObjectStore() {
+    final transaction = _database.transaction(
+      storeName,
+      mode: IDBTransactionMode.readwrite,
+    );
+    final store = transaction.objectStore(storeName);
+    return store;
+  }
+
+  @override
+  Future<void> addItem(String string, String timestamp) async {
+    _currentTotalByteSize += utf8.encode(string).length;
+    _currentTotalByteSize += utf8.encode(timestamp).length;
+    _currentTotalByteSize += 8; // 8 bytes for the id
+    await _databaseOpenEvent;
+    await _getObjectStore().add(string, timestamp).future;
+  }
+
+  @override
+  Future<Iterable<QueuedItem>> getCount([int? count]) async {
+    final readValues = <QueuedItem>[];
+
+    await _databaseOpenEvent;
+    final store = _getObjectStore();
+    final request = store.getAll(null, count);
+
+    await request.future;
+
+    for (final elem in request.result) {
+      final value = elem as Object;
+      final id = getProperty<int>(value, 'id');
+      final string = getProperty<String>(value, 'value');
+      final timestamp = getProperty<String>(value, 'timestamp');
+      readValues.add(
+        QueuedItem(id: id, value: string, timestamp: timestamp),
+      );
+    }
+    return readValues;
+  }
+
+  @override
+  Future<void> deleteItems(Iterable<QueuedItem> items) async {
+    if (items.isEmpty) return;
+
+    _currentTotalByteSize -= utf8.encode(items.first.value).length;
+    _currentTotalByteSize -= utf8.encode(items.first.timestamp).length;
+    _currentTotalByteSize -= 8; // 8 bytes for the id
```

Comment:
Needs a TODO here and above. Running `utf8.encode` twice for every log entry is a poor use of CPU cycles and should not be done on the main thread. Improvement options are:
- Offload to another thread with worker bee (several examples in the repo)
- Use `String.length` instead, which gives the size in UTF-16 code points

It's also not clear to me if this is even correct. Here's a screenshot from Firefox showing how it stores a JSON blob in IndexedDB.

<img width="1377" alt="Screenshot 2023-08-21 at 4 25 38 PM" src="https://github.com/aws-amplify/amplify-flutter/assets/24740863/8b1de73f-368c-41c7-a2f6-0ab3e382ddfd">